File: packages/agent/src/tools/read-file.ts
import { readFileSync } from "node:fs";
import { z } from "zod";
import type { ToolDefinition } from "./types.js";

const schema = z.object({
  path: z.string().describe("Absolute or relative path to the file to read"),
});

export const readFileTool: ToolDefinition = {
  name: "read_file",
  description: "Read the contents of a file. Returns file content with line numbers.",
  inputSchema: schema,
  async execute(args) {
    const { path } = schema.parse(args);
    try {
      const content = readFileSync(path, "utf-8");
      const numbered = content
        .split("\n")
        .map((line, i) => `${String(i + 1).padStart(4)} | ${line}`)
        .join("\n");
      return { output: numbered };
    } catch (err) {
      return { output: `Error reading file: ${(err as Error).message}`, isError: true };
    }
  },
};

File: crates/oni-agent/src/tools/read_file.rs
use super::Tool;
use oni_core::error::Result;
use oni_core::types::ToolCapability;

pub struct ReadFileTool;

impl Tool for ReadFileTool {
    fn name(&self) -> &str {
        "read_file"
    }

    fn required_capabilities(&self) -> &[ToolCapability] {
        &[ToolCapability::ReadFs]
    }

    fn description(&self) -> &str {
        "Read the contents of a file at the given path"
    }

    fn schema(&self) -> serde_json::Value {
        serde_json::json!({
            "type": "function",
            "function": {
                "name": "read_file",
                "description": self.description(),
                "parameters": {
                    "type": "object",
                    "properties": {
                        "path": {
                            "type": "string",
                            "description": "Absolute or relative path to the file"
                        }
                    },
                    "required": ["path"]
                }
            }
        })
    }

    fn execute(&self, args: serde_json::Value) -> Result<String> {
        let path = args
            .get("path")
            .and_then(|v| v.as_str())
            .ok_or_else(|| oni_core::error::err!("Missing 'path' argument"))?;

        match std::fs::read_to_string(path) {
            Ok(content) => {
                // Truncate very large files
                if content.len() > 100_000 {
                    let end = {
                        let mut e = 100_000_usize.min(content.len());
                        while e > 0 && !content.is_char_boundary(e) { e -= 1; }
                        e
                    };
                    Ok(format!(
                        "{}...\n\n[Truncated: file is {} bytes]",
                        &content[..end],
                        content.len()
                    ))
                } else {
                    Ok(content)
                }
            }
            Err(e) => Ok(format!("Error reading file '{}': {}", path, e)),
        }
    }
}
